test-data/unit/check-namedtuple.test

```diff
@@ -185,3 +200,104 @@ class C(B): pass
 B(1).b
 C(2).b
 [builtins fixtures/property.pyi]
+
+[case testNamedTupleAsDict]
+from collections import namedtuple, OrderedDict
+
+X = namedtuple('X', ['x', 'y'])
+x = None  # type: X
+a = None  # type: OrderedDict
+a = x._asdict()
+
+[case testNamedTupleReplace]
+from collections import namedtuple
+
+X = namedtuple('X', ['x', 'y'])
+x = None  # type: X
+x = x._replace()
+x = x._replace(y=5)
+x = x._replace(x=3)
+x = x._replace(x=3, y=5)
+x = x._replace(z=5)  # E: Unexpected keyword argument "z" for X._replace
+x = x._replace(5)  # E: Too many positional arguments for X._replace
+
+[case testNamedTupleReplaceAsClass]
+from collections import namedtuple
+
+X = namedtuple('X', ['x', 'y'])
+x = None  # type: X
+X._replace(x, x=1, y=2)
+X._replace(x=1, y=2)  # E: Missing positional argument "self" in call to X._replace
+
+[case testNamedTupleReplaceTyped]
+from typing import NamedTuple
+
+X = NamedTuple('X', [('x', int), ('y', str)])
+x = None  # type: X
+x = x._replace(x=5)
+x = x._replace(y=5)  # E: Argument 1 to X._replace has incompatible type "int"; expected "str"
+
+
+[case testNamedTupleMake]
+from typing import NamedTuple
+
+X = NamedTuple('X', [('x', int), ('y', str)])
+x = None  # type: X
+x = X._make([5, 'a'])
+
+[builtins fixtures/list.pyi]
+
+[case testNamedTupleFields]
+from typing import NamedTuple
+
+X = NamedTuple('X', [('x', int), ('y', str)])
+a = None  # type: str
+b = None  # type: str
+a, b = X._fields
+x = None  # type: X
+a, b = x._fields
+c = None  # type: int
+a, c = X._fields  # E: Incompatible types in assignment (expression has type "str", variable has type "int")
+
+[case testNamedTupleSource]
+from typing import NamedTuple
+
+X = NamedTuple('X', [('x', int), ('y', str)])
+x = None  # type: X
+a = None  # type: str
+b = None  # type: int
+a = X._source
+b = X._source  # E: Incompatible types in assignment (expression has type "str", variable has type "int")
+
+[case testNamedTupleUnit]
+from typing import NamedTuple
+
+X = NamedTuple('X', [])
+x = X()  # type: X
+x = x._replace()
+_ = x._fields[0]  # E: Tuple index out of range
+
+[case testNamedTupleJoinNamedTuple]
+from typing import NamedTuple
+
+X = NamedTuple('X', [('x', int), ('y', str)])
+Y = NamedTuple('Y', [('x', int), ('y', str)])
+xy = None
+if None:
+    xy = X(3, 'b')
```

Comment:
That has been special cased heavily so it's not just a join.
On Wed, Aug 31, 2016 at 22:29 Elazar <notifications@github.com> wrote:

> In test-data/unit/check-namedtuple.test
> <https://github.com/python/mypy/pull/1810#discussion_r77076682>:
>
> > +[case testNamedTupleUnit]
> > +from typing import NamedTuple
> > +
> > +X = NamedTuple('X', [])
> > +x = X()  # type: X
> > +x = x._replace()
> > +_ = x._fields[0]  # E: Tuple index out of range
> > +
> > +[case testNamedTupleJoinNamedTuple]
> > +from typing import NamedTuple
> > +
> > +X = NamedTuple('X', [('x', int), ('y', str)])
> > +Y = NamedTuple('Y', [('x', int), ('y', str)])
> > +xy = None
> > +if None:
> > +    xy = X(3, 'b')
>
> What it performs then? I'd expect A if _ else B to be join(A, B).
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/python/mypy/pull/1810/files/71f0d781fcfaf01c2dd51185d016f17a0643e7a4#r77076682>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABDnx02iEtlnngYuwv5s5UDlno_0tLhXks5qlfGHgaJpZM4JF86j>
> .
>